dspy/experimental/module_graph.py

```diff
@@ -0,0 +1,127 @@
+            if key in str(module_type):
+                return base_name + suffix
+        return base_name
+
+    def process_submodules(self, module_name, module):
+        """ Process submodules of a module and add them to the graph"""
+
+        for sub_module_name, sub_module in module.__dict__.items():
+            if isinstance(sub_module, dspy.Predict):
+                sub_module_name = self.generate_module_name(sub_module_name, type(sub_module))
+                self.process_submodule(f"{module_name}__{sub_module_name}", sub_module)
+            elif isinstance(sub_module, (dspy.Module, dspy.Retrieve)):
+                self.add_module(f"{module_name}__{sub_module_name}", sub_module)
+                if isinstance(sub_module, dspy.Retrieve):
+                    self.graph.edge("rm", 'lm', label='RM used in Module')
+
+    def process_submodule(self, sub_module_name, sub_module):
+        """Process a submodule and add it to the graph"""
+
+        for field_type, fields in [('input', sub_module.signature.input_fields),
+                                   ('output', sub_module.signature.output_fields)]:
+            for field_name, field in fields.items():
+                node_id = f"{sub_module_name}_{field_type}_{field_name}"
+                if node_id not in self.nodes:
+                    label = f"{field_name}: ({field.json_schema_extra['desc']})"
+                    self.graph.node(node_id, label=label, shape='ellipse')
+                    self.nodes.add(node_id)
+                edge_direction = (node_id, sub_module_name) if field_type == 'input' else (sub_module_name, node_id)
+                self.graph.edge(*edge_direction)
+
+        # Add node for the submodule itself
+        self.graph.node(sub_module_name, label=sub_module_name, shape='box')
+        self.nodes.add(sub_module_name)
+
+        # Connect submodule to LM if configured
+        if 'lm' in self.nodes:
+            self.graph.edge('lm', sub_module_name, label='LM used in Module')
+
+    def render_graph(self, filename=None):
+        """Render the graph to a file(png)"""
+        if filename is None:
+            filename = self.module_name
+        self.graph.render(filename)
+
+
+
+# Example usage of the ModuleGraph class:
+
+# class GenerateAnswer(dspy.Signature):
+#   "Answer with long and detailled answers"
+#   context = dspy.InputField(desc="may content relevant facts")
+#   question = dspy.InputField()
+#   answer = dspy.OutputField(desc="often between 10 and 50 words")
+
+# class RAG(dspy.Module):
+#   def __init__(self, num_passages=3):
+#     super().__init__()
+#     self.retrieve = dspy.Retrieve(k=num_passages)
+#     self.generate_answer = dspy.ChainOfThought(GenerateAnswer)
+
+#   def forward(self, question):
+#     context = self.retrieve(question).passages
+#     prediction = self.generate_answer(context=context, question=question)
+#     return dspy.Prediction(context=context, answer=prediction.answer)
+
+# rag_system = RAG()
```

Comment:
would be good to add the configurations of the LM/RM in DSPy for accurate reproducibility

ex:
```dspy.settings.configure(lm=turbo, rm=colbertv2_wiki17_abstracts)`

Reply:
Sure, no problem.